### PR TITLE
Daily add keys per entity account count

### PR DESCRIPTION
**Pull request by jschintz**
Count of # of add key events per entity account that added the key.
Measures # of times entities have added users to their app/nft/guild.

### CODE REVIEW

File: aggregations/db_tables/daily_new_accounts_by_app_count.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+            FROM public.near_ecosystem_entities e, unnest(string_to_array(e.contract, ', ')) s(token)
+            WHERE category LIKE '%%app%%' AND length(contract)>0
+        '''
+
+        with self.analytics_connection.cursor() as analytics_cursor:
+            # Get all contracts that were added before our time range
+            analytics_cursor.execute(all_apps)
+            app_contracts = analytics_cursor.fetchall()
+
+        #concat table values into case statement
```
Comment (frol) [inline aggregations/db_tables/daily_new_accounts_by_app_count.py:46]: It would be great if comments would actually help

File: aggregations/db_tables/daily_new_accounts_by_app_count.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+from ..periodic_aggregations import PeriodicAggregations
+
+
+class DailyNewAccountsByAppCount(PeriodicAggregations):
+    @property
+    def sql_create_table(self):
+        # Suppose we have at most 10^5 (100K) transactions per second.
+        # In the worst case, they are all from one account.
+        # It gives ~10^10 transactions per day.
+        # It means we fit into BIGINT (10^18)
```
Comment (frol) [inline aggregations/db_tables/daily_new_accounts_by_app_count.py:13]: You seemed to miss the point of the comment and just left it here.

File: aggregations/db_tables/daily_new_accounts_by_app_count.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    def sql_select(self):
+        # retrieve app data from near- analytics
+        all_apps = '''
+            SELECT token, slug as app_id
+            FROM public.near_ecosystem_entities e, unnest(string_to_array(e.contract, ', ')) s(token)
+            WHERE category LIKE '%%app%%' AND length(contract)>0
+        '''
+
+        with self.analytics_connection.cursor() as analytics_cursor:
+            # Get all contracts that were added before our time range
```
Comment (frol) [inline aggregations/db_tables/daily_new_accounts_by_app_count.py:42]: I don't get the part about the time range

File: aggregations/db_tables/daily_new_accounts_by_app_count.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
+import datetime
+
+from . import DAY_LEN_SECONDS, daily_start_of_range, time_json
+from ..periodic_aggregations import PeriodicAggregations
+
+
+class DailyNewAccountsByAppCount(PeriodicAggregations):
```
Comment (frol) [inline aggregations/db_tables/daily_new_accounts_by_app_count.py:7]: When you depend on some other table from near analytics database, you should define `dependencies` method like in the example I gave you: https://github.com/near/near-analytics/blob/main/aggregations/db_tables/daily_new_unique_contracts_count.py#L9-L10

File: aggregations/base_aggregations.py [hunk 5]
```diff
@@ -34,4 +34,4 @@ def drop_table(self):
     # Be careful, don't create circular dependencies
     @abc.abstractmethod
     def dependencies(self) -> list:
-        pass
+        pass
```
Comment (frol) [inline aggregations/base_aggregations.py:37]: Don't commit unnecessary changes

**Review by frol — CHANGES_REQUESTED** [review]

File: aggregations/db_tables/daily_new_accounts_by_app_count.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+            WHERE app = TRUE AND length(contract)>0
+        '''
+
+        with self.analytics_connection.cursor() as analytics_cursor:
+            analytics_cursor.execute(all_apps)
+            app_contracts = analytics_cursor.fetchall()
+
+        string = ''
+        for index, tuple in enumerate(app_contracts):
+#            contract = ''
```
Comment (frol) [inline aggregations/db_tables/daily_new_accounts_by_app_count.py:45]: Clean up

File: aggregations/db_tables/daily_new_accounts_by_app_count.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+            contract = str(tuple[0])
+            app = str(tuple[1])
```
Comment (frol) [inline aggregations/db_tables/daily_new_accounts_by_app_count.py:47]: Aren't those already strings? Why do we need to `str()` those values?

File: aggregations/db_tables/daily_new_accounts_by_app_count.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        return ['near_ecosystem_entities']
+
+    @property
+    def sql_create_table(self):
+        return '''
+            CREATE TABLE IF NOT EXISTS daily_new_accounts_by_app_count
+            (
+                collected_for_day  DATE NOT NULL,
+                app_id TEXT NOT NULL,
+                new_accounts INTEGER NOT NULL,
```
Comment (frol) [inline aggregations/db_tables/daily_new_accounts_by_app_count.py:18]: Let's rename it to `new_accounts_count`

File: aggregations/db_tables/daily_new_accounts_by_app_count.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    @property
+    def sql_drop_table(self):
+        return '''
+            DROP TABLE IF EXISTS daily_new_accounts_by_app_count
+        '''
+
+    @property
+    def sql_select(self):
+        all_apps = '''
+            SELECT token, slug as app_id
```
Comment (frol) [inline aggregations/db_tables/daily_new_accounts_by_app_count.py:34]: 1. What is the meaning for `token`?
2. Given you renamed `slug` to `app_id`, I feel that the original table might have the column named as `app_id` then rather than `slug`. What do you think?

**Review by frol — CHANGES_REQUESTED** [review]
Resolve merge conflicts (merge main branch into your branch)

File: aggregations/db_tables/daily_new_accounts_by_app_count.py (hunk 9 above)
Comment (jschintz) [reply aggregations/db_tables/daily_new_accounts_by_app_count.py]: Token is a keyword driving functionality of sql being used on the join.  I renamed to contract.  But it is being referenced by position further down in the code, so the name doesn't matter, the position does.  

Slug is the name used by the ecosystem team.  It is where you can find the "entity"  The page URL that follows after near.org/ecosystem/  .  This will be used for apps, nft & guilds I believe, so let's  leave that generalized in the entity table.

File: aggregations/db_tables/daily_new_accounts_by_app_count.py (hunk 7 above)
Comment (jschintz) [reply aggregations/db_tables/daily_new_accounts_by_app_count.py]: removed

File: aggregations/db_tables/daily_new_accounts_by_app_count.py (hunk 6 above)
Comment (jschintz) [reply aggregations/db_tables/daily_new_accounts_by_app_count.py]: removed

**Comment by jschintz** [comment]
@frol & I added in a bit of a re-work for getting new accounts, creating a dependency to preaggregate the data.  The earlier sql was not performant, but this adds more # records to the analytics-near db.  Let's chat if you need this reworked.

File: aggregations/db_tables/daily_add_key_by_app_count.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+            DROP TABLE IF EXISTS daily_add_key_by_app_count
+        """
+
+    @property
+    def sql_select(self):
+        all_apps = """
+            SELECT token as contract_id, 
+            slug as app_id
+            FROM public.near_ecosystem_entities e, unnest(string_to_array(e.contract, ', ')) s(token)
+            WHERE app = TRUE AND length(contract)>0
```
Comment (telezhnaya) [inline aggregations/db_tables/daily_add_key_by_app_count.py:38]: It is super hard to read. This construction
```
unnest(string_to_array(e.contract, ', ')) s(token)
```
could not appear if you design `near_ecosystem_entities` properly.

`near_ecosystem_entities.contract` has `text` type and it is named in a single form, why do we try to split it by comma?

Comment (telezhnaya) [reply aggregations/db_tables/daily_add_key_by_app_count.py:38]: It's the first time when I see `s(token)` construction. Can you please explain what does it do? What's the difference between your usual way to create the alias (`as token`) and this way to create the alias?

Comment (telezhnaya) [reply aggregations/db_tables/daily_add_key_by_app_count.py:38]: `SELECT token as contract_id` confuses me a lot.
You created `unnest(string_to_array(e.contract, ', ')) s(token)` with `token` name, then you renamed it in the same expression by `token as contract_id`, why do you use so complicated logic?

I also have some domain-specific questions. What is the token and what is the contract id? Why do we say that token and contract_id are the same things?

File: aggregations/db_tables/daily_add_key_by_app_count.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+        """
+
+        with self.analytics_connection.cursor() as analytics_cursor:
+            analytics_cursor.execute(all_apps)
+            app_contracts = analytics_cursor.fetchall()
```
Comment (telezhnaya) [inline aggregations/db_tables/daily_add_key_by_app_count.py:43]: While Python allows creating variable in a block of code and using it after we quit this block, I feel it's not the best practice and it makes the reading more complicated. 
I suggest you to move this block to a function with `app_contracts` as the returned value.

File: aggregations/db_tables/daily_add_key_by_app_count.py [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+
+        string = ""
```
Comment (telezhnaya) [inline aggregations/db_tables/daily_add_key_by_app_count.py:45]: Confusing naming

File: aggregations/db_tables/daily_add_key_by_app_count.py [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+        receipt_receiver_account_id,
+        count(*) as new_accounts_count
+        FROM public.action_receipt_actions a
+        WHERE action_kind IN ('ADD_KEY')
+            AND args ->'access_key' -> 'permission' ->> 'permission_kind' = 'FUNCTION_CALL'
+            AND args ->'access_key' -> 'permission' -> 'permission_details' ->> 'receiver_id' != receipt_receiver_account_id 
+            AND args ->'access_key' -> 'permission' -> 'permission_details' ->> 'receiver_id' != 'near'
+            AND receipt_included_in_block_timestamp  >= %(from_timestamp)s
+            AND receipt_included_in_block_timestamp  < %(to_timestamp)s
+        GROUP BY 1,2
```
Comment (telezhnaya) [inline aggregations/db_tables/daily_add_key_by_app_count.py:68]: Have you tried to compose this query manually and measure the performance?

File: aggregations/db_tables/daily_new_accounts_by_app_count.py [hunk 14]
```diff
@@ -0,0 +1,80 @@
+import datetime
```
Comment (telezhnaya) [inline aggregations/db_tables/daily_new_accounts_by_app_count.py:1]: If you need to add 2 separate statistics, it's better to create 2 separate reviews for that, it simplifies (read: makes faster) the review process

**Review by telezhnaya — CHANGES_REQUESTED** [review]
Could you please explain the reasons why do we need such tables and how do we plan to use them?
Do we need `daily_add_key_by_app_count` only for collecting data for `daily_new_accounts_by_app_count`?
Have you considered other ways to collect the same data?

File: aggregations/db_tables/daily_add_key_by_app_count.py (hunk 12 above)
Comment (jschintz) [reply aggregations/db_tables/daily_add_key_by_app_count.py:45]: What would you like it named?

File: aggregations/db_tables/daily_add_key_by_app_count.py (hunk 10 above)
Comment (jschintz) [reply aggregations/db_tables/daily_add_key_by_app_count.py:38]: You and Frol are correct, I can remove token keyword.  That is only necessary in the MSSQL dialect.  I just updated.

I agree, near_ecosystem_entities could have a better layout.  That is the data being provided today.  
It is the data provided from the near/ecosystems repository without alteration.  For now this is the best data that is available.  In general, I attempt to create an extract-load-transform where the data gets stored before I do transformations.  

These naming conventions like "slug", "contract", ect...  come from that near/ecosystems repository.  Please feel free to suggest any name changes, this will be your baby long-term.

**Comment by jschintz** [comment]
> Could you please explain the reasons why do we need such tables and how do we plan to use them? Do we need `daily_add_key_by_app_count` only for collecting data for `daily_new_accounts_by_app_count`? Have you considered other ways to collect the same data?

We are only building these tables for one purpose, supporting Eric's dashboard.  

Yes, daily_add_key_by_app_count is only being used by us to create daily_new_accounts_by_app_count.  I tested a bunch of different approaches, CTE's, subqueries, getting Neil to just use the daily_add_key_by_app_count  table directly. 
Checkout the performance, these queries are speedy!  

It could have other uses though ...  # of users who have repeat account key adds.  Velocity of account key adds, anything to do with account key adds.

My opinion though, move all of this to big query!

File: aggregations/db_tables/daily_new_accounts_by_app_count.py (hunk 14 above)
Comment (jschintz) [reply aggregations/db_tables/daily_new_accounts_by_app_count.py:1]: Yeah, I totally get that for the future.  Sorry, this grew to 2 at the end of deployment.  The queries would run otherwise!

Neil needed something to test against.  Maybe this needs a dev server long-term?

File: aggregations/db_tables/daily_add_key_by_app_count.py (hunk 13 above)
Comment (jschintz) [reply aggregations/db_tables/daily_add_key_by_app_count.py:68]: Yes, I did extensive testing.  This was the quickest solution I found, even with the dependency.

File: aggregations/db_tables/daily_add_key_by_app_count.py (hunk 11 above)
Comment (jschintz) [reply aggregations/db_tables/daily_add_key_by_app_count.py:43]: I'll do you one even better...  This whole thing it overly complicated &  SQL injection is terrible for analytics.  I did not want to built it this way.

What do you think about pushing the call to  table near_ecosystem_entities to the other job 'daily_new_accounts_by_app_count'?

File: aggregations/db_tables/daily_add_key_by_app_count.py (hunk 13 above)
Comment (telezhnaya) [reply aggregations/db_tables/daily_add_key_by_app_count.py:68]: Could you please share the details of the testing? And the query that you've composed

**Comment by frol** [comment]
> Checkout the performance, these queries are speedy!

It would have been more helpful if you could be more specific, i.e. "these queries take around 10 seconds to execute on mainnet and 3 seconds on testnet".

> My opinion though, move all of this to big query!

Can you share the exact reasons, please? Our analytics database is quite tiny compared to 800+GB of Indexer for Explorer database, so I don't see any good reason to migrate analytics database to big query.

**Comment by jschintz** [comment]
For the CSV:  In general, I would advocate for an ELT over ETL
architecture.  THis clearly isn't happening with NEAR-Analytics anyways,
but there are many benefits to keep all your data, in one spot.  Having
this data directly inside NEAR-Analytics would give the following benefits.

   - Data extraction & load would only happen once.  Today this csv is only
   be used for apps, but it will grow to guilds, nft's and who knows what
   else.  By keeping your data in one location you can begin to migrate
   towards a concept of a single source of truth.
   - Databases are more efficient at processing data than python because
   db's specialize at that.   By having this in near-analytics, you will get
   better performance through the use of indexes.
   - You aren't selecting 'All' the records, its a subset of apps, nfts,
   guilds.

For receipt_receiver_account_id this is the name of the account issuing the
request to add a key.

**Comment by jschintz** [comment]
> > Checkout the performance, these queries are speedy!
> 
> It would have been more helpful if you could be more specific, i.e. "these queries take around 10 seconds to execute on mainnet and 3 seconds on testnet".

Queries range from 2-5 seconds since august time frame on mainnet.  
> 
> > My opinion though, move all of this to big query!
> 
> Can you share the exact reasons, please? Our analytics database is quite tiny compared to 800+GB of Indexer for Explorer database, so I don't see any good reason to migrate analytics database to big query.

I was referring to getting the indexer directly into a columnar store db and using that for the analytics database.  Separating the write from read would give you substantial benefits, and its a lot of overhead to learn near-analytics python repository.   Long-term doing the aggregations metric by metric with an ETL approach wont be sustainable.  It took too much time was invested just to get 2 metrics,

**Comment by frol** [comment]
> For receipt_receiver_account_id this is the name of the account issuing the request to add a key.

Given that there is `collected_for_day` & `app_id` & `app_user_account_id` & `add_key_count`, in plain English it sounds like you wanted to count how many times some user logged into a given application in a given day. This makes sense, but it took me quite some time to deduce from the code just because the naming does not help me at all.

> For the CSV:  In general, I would advocate for an ELT over ETL architecture.  THis clearly isn't happening with NEAR-Analytics anyways, but there are many benefits to keep all your data, in one spot. 

Well, at least I now see where was the disconnect. You tried to enforce ELT while we expect it to be ETL. NEAR-Analytics is the transform layer, and we build ETL here and expect to load (store) only structured data while you tried to put raw data to the database. I feel ELT is a waste of resources since I don't see a difference of processing a CSV/JSON file or same unstructured data from my own data source.

>   - Data extraction & load would only happen once. 

That is a weird assumption given that we depend on some external data source which gets updated, and you only created one extra transform step and complicated your own life.

>    - Databases are more efficient at processing data than python because db's specialize at that.
>    - You aren't selecting 'All' the records, its a subset of apps, nfts, guilds.

So you are saying that loading the whole CSV/JSON file into the database + selecting a subset of it is faster than selecting the relevant records from CSV/JSON file, is this what you mean?

> Queries range from 2-5 seconds since august time frame on mainnet.

:heart: 

> I was referring to getting the indexer directly into a columnar store db and using that for the analytics database. Separating the write from read would give you substantial benefits, and its a lot of overhead to learn near-analytics python repository. Long-term doing the aggregations metric by metric with an ETL approach wont be sustainable.

Well, we have that on our roadmap for Q1.

> It took too much time was invested just to get 2 metrics,

It took us a few days to build the whole framework and collect all the other metrics we have in this repo :shrug:

**Review by frol — CHANGES_REQUESTED** [review]

**Comment by frol** [comment]
This PR was superseeded by #39.
